nalgebra-sparse/src/io/matrix_market.rs

```diff
@@ -0,0 +1,1293 @@
+                            r + 1
+                        ),
+                    ));
+                }
+                // don't need to add twice if the element in on diagonal
+                if r != c {
+                    rows.push(c);
+                    cols.push(r);
+                    data.push(d.conjugate()?);
+                }
+            }
+        }
+    }
+    Ok(CooMatrix::try_from_triplets(
+        shape.0, shape.1, rows, cols, data,
+    )?)
+}
+
+#[inline]
+/// do a quick check it the entry is in the lower triangle part of the matrix
+fn check_lower_triangle(r: usize, c: usize) -> Result<(), MatrixMarketError> {
+    if c > r {
+        return Err(MatrixMarketError::from_kind_and_message(
+            MatrixMarketErrorKind::NotLowerTriangle,
+            format!(
+                "Entry: row {} col {} should be put into lower triangle",
+                r, c
+            ),
+        ));
+    }
+    Ok(())
+}
+
+#[inline]
+/// Parse a pest structure to a Typecode of the matrix.
+fn parse_header(inner: &mut Pairs<'_, Rule>) -> Typecode {
+    // unwrap() in this function are guaranteed by parsing the data
+    Typecode {
+        sparsity: inner
+            .next()
+            .unwrap()
+            .as_str()
+            .to_ascii_lowercase()
+            .parse::<Sparsity>()
+            .unwrap(),
+        datatype: inner
+            .next()
+            .unwrap()
+            .as_str()
+            .to_ascii_lowercase()
+            .parse::<DataType>()
+            .unwrap(),
+        storagescheme: inner
+            .next()
+            .unwrap()
+            .as_str()
+            .to_ascii_lowercase()
+            .parse::<StorageScheme>()
+            .unwrap(),
+    }
+}
+
+// Parse shape starts here-------------------------------------------------
+
+/// Parse a pest structure to sparse shape information, including 3 int, which are number of rols, cols and non-zeros.
+fn parse_sparse_shape(
+    inner: &mut Pairs<'_, Rule>,
+    storagescheme: &StorageScheme,
+) -> Result<(usize, usize, usize), MatrixMarketError> {
+    // unwrap() in this function are guaranteed by parsing the data
+    let shape_inner = inner.next().unwrap();
+    if shape_inner.as_rule() != Rule::SparseShape {
+        return Err(MatrixMarketError::from_kind_and_message(MatrixMarketErrorKind::ParsingError,format!("
+        Shape shape line requires 3 int numbers as number of rows, columns and non-zeros, but line {} was provided here.
+        ",shape_inner.as_str())));
+    }
+
+    let mut inner = shape_inner.into_inner();
+
+    let r = inner.next().unwrap().as_str().parse::<usize>().unwrap();
+    let c = inner.next().unwrap().as_str().parse::<usize>().unwrap();
+    let nnz = inner.next().unwrap().as_str().parse::<usize>().unwrap();
+
+    // shape information can't use 0 as dimension
+    if r * c * nnz == 0 {
```

Comment:
Does this mean we cannot load an empty matrix?

Also, should check the numbers individually in order to avoid possible overflow.